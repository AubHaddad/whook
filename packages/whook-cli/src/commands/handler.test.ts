import { jest } from '@jest/globals';
import initHandlerCommand from './handler.js';
import { YError } from 'yerror';
import type { LogService } from 'common-services';
import type { PromptArgs } from '../services/promptArgs.js';
import type { Injector } from 'knifecycle';

describe('handlerCommand', () => {
  const promptArgs = jest.fn<PromptArgs>();
  const log = jest.fn<LogService>();
  const $injector = jest.fn<Injector<any>>();

  beforeEach(() => {
    promptArgs.mockReset();
    log.mockReset();
    $injector.mockReset();
  });

  describe('should work', () => {
    it('with all parameters', async () => {
      $injector.mockResolvedValueOnce({
        putEcho: async ({ body }) => ({
          status: 200,
          body,
        }),
      });
      promptArgs.mockResolvedValueOnce({
        command: 'whook',
        rest: ['handler'],
        namedArguments: {
          name: 'putEcho',
          parameters: '{"body": {"echo": "YOLO!"} }',
        },
      });

      const handlerCommand = await initHandlerCommand({
        promptArgs,
        log,
        $injector,
      });

      await handlerCommand();

      expect({
        promptArgsCalls: promptArgs.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        injectorCalls: $injector.mock.calls,
      }).toMatchSnapshot();
    });

    it('with handler only', async () => {
      $injector.mockResolvedValueOnce({
        getPing: async ({ body }) => ({
          status: 200,
          body,
        }),
      });
      promptArgs.mockResolvedValueOnce({
        command: 'whook',
        rest: ['handler'],
        namedArguments: {
          name: 'getPing',
        },
      });

      const handlerCommand = await initHandlerCommand({
        promptArgs,
        log,
        $injector,
      });

      await handlerCommand();

      expect({
        promptArgsCalls: promptArgs.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        injectorCalls: $injector.mock.calls,
      }).toMatchSnapshot();
    });
  });

  describe('should fail', () => {
    it('with non JSON parameters', async () => {
      $injector.mockResolvedValueOnce({
        putEcho: async ({ body }) => ({
          status: 200,
          body,
        }),
      });
      promptArgs.mockResolvedValueOnce({
        command: 'whook',
        rest: ['handler'],
        namedArguments: {
          name: 'putEcho',
          parameters: '{"body: {"echo": "YOLO!"} }',
        },
      });

      const handlerCommand = await initHandlerCommand({
        promptArgs,
        log,
        $injector,
      });

      try {
        await handlerCommand();
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: (err as YError).code,
          errorParams: (err as YError).params,
          promptArgsCalls: promptArgs.mock.calls,
          logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
          injectorCalls: $injector.mock.calls,
        }).toMatchSnapshot();
      }
    });

    it('with a failing handler', async () => {
      $injector.mockResolvedValueOnce({
        putEcho: async () => {
          throw new YError('E_HANDLER_ERROR');
        },
      });
      promptArgs.mockResolvedValueOnce({
        command: 'whook',
        rest: ['handler'],
        namedArguments: {
          name: 'putEcho',
          parameters: '{"body": {"echo": "YOLO!"} }',
        },
      });

      const handlerCommand = await initHandlerCommand({
        promptArgs,
        log,
        $injector,
      });

      try {
        await handlerCommand();
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: (err as YError).code,
          errorParams: (err as YError).params,
          promptArgsCalls: promptArgs.mock.calls,
          logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
          injectorCalls: $injector.mock.calls,
        }).toMatchSnapshot();
      }
    });
  });
});
