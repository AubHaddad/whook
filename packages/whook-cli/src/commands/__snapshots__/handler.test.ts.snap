// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handlerCommand should fail with a failing handler 1`] = `
{
  "errorCode": "E_UNEXPECTED_SUCCESS",
  "errorParams": [],
  "injectorCalls": [
    [
      [
        "putEcho",
      ],
    ],
  ],
  "logCalls": [
    [
      "debug",
      "handler",
      "putEcho",
    ],
    [
      "debug",
      "parameters",
      {
        "body": {
          "echo": "YOLO!",
        },
      },
    ],
    [
      "error",
      "Got an error while running the handler.",
    ],
  ],
  "promptArgsCalls": [
    [],
  ],
}
`;

exports[`handlerCommand should fail with non JSON parameters 1`] = `
{
  "errorCode": "E_BAD_PARAMETERS",
  "errorParams": [
    "{"body: {"echo": "YOLO!"} }",
    "Unexpected token e in JSON at position 10",
  ],
  "injectorCalls": [],
  "logCalls": [],
  "promptArgsCalls": [
    [],
  ],
}
`;

exports[`handlerCommand should work with all parameters 1`] = `
{
  "injectorCalls": [
    [
      [
        "putEcho",
      ],
    ],
  ],
  "logCalls": [
    [
      "debug",
      "handler",
      "putEcho",
    ],
    [
      "debug",
      "parameters",
      {
        "body": {
          "echo": "YOLO!",
        },
      },
    ],
    [
      "info",
      "{
  "status": 200,
  "body": {
    "echo": "YOLO!"
  }
}",
    ],
  ],
  "promptArgsCalls": [
    [],
  ],
}
`;

exports[`handlerCommand should work with handler only 1`] = `
{
  "injectorCalls": [
    [
      [
        "getPing",
      ],
    ],
  ],
  "logCalls": [
    [
      "debug",
      "handler",
      "getPing",
    ],
    [
      "debug",
      "parameters",
      {},
    ],
    [
      "info",
      "{
  "status": 200
}",
    ],
  ],
  "promptArgsCalls": [
    [],
  ],
}
`;
